gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -1494,65 +1498,72 @@ private Credential getCredential(Configuration config)
 
     // If impersonation service account exists, then use current credential to request access token
     // for the impersonating service account.
-    return getImpersonatedCredential(config, credential).orElse(credential);
+    return getImpersonatedCredential(config, gcsFsOptions, credential).orElse(credential);
   }
 
   /**
    * Generate a {@link Credential} from the internal access token provider based on the service
    * account to impersonate.
    */
   private static Optional<Credential> getImpersonatedCredential(
-      Configuration config, Credential credential) throws IOException {
-    GoogleCloudStorageOptions options =
-        GoogleHadoopFileSystemConfiguration.getGcsFsOptionsBuilder(config)
-            .build()
-            .getCloudStorageOptions();
-
-    String serviceAccountToImpersonate =
-        IMPERSONATION_SERVICE_ACCOUNT_SUFFIX
-            .withPrefixes(CONFIG_KEY_PREFIXES)
-            .get(config, config::get);
-    if (isNullOrEmpty(serviceAccountToImpersonate)) {
-      Optional<String> serviceAccountToImpersonateFromUser =
-          getMatchedServiceAccountToImpersonateFromGroup(
-              options.getUserImpersonationServiceAccounts(),
-              ImmutableList.of(UserGroupInformation.getCurrentUser().getShortUserName()));
-      Optional<String> serviceAccountToImpersonateFromGroup =
-          getMatchedServiceAccountToImpersonateFromGroup(
-              options.getGroupImpersonationServiceAccounts(),
-              ImmutableList.copyOf(UserGroupInformation.getCurrentUser().getGroupNames()));
-      serviceAccountToImpersonate =
-          serviceAccountToImpersonateFromUser.orElse(
-              serviceAccountToImpersonateFromGroup.orElse(null));
+      Configuration config, GoogleCloudStorageFileSystemOptions gcsFsOptions, Credential credential)
```

Comment:
I thought designated service account will always take precedence?

Reply:
I think that it's the best if the narrower scope impersonation configuration takes precedence over broader-scope impersonation. I.e user impersonation overrides group impersonation which in turn overrides default/generic impersonation.

Is this what you thought too?

Reply:
Agreed with narrower scope should take precedence. The original design was if a single SA is set up thru IMPERSONATION_SERVICE_ACCOUNT_SUFFIX this will take precedence over everything(I somehow have such impression that you also agreed on this in one of the comments you had on my PR).

But I guess this is not a critical order anyway, since I would doubt it happens often that properties of both two categories will be set at the same time. It's helpful to document the precedence explicitly tho.

Reply:
If we keep like this way maybe add something like "Note that if any of these two properties is set, it will override the value of
`fs.gs.auth.impersonation.service.account` if set." down below the Service account impersonation section.

Reply:
Done.